### DISCUSSION TITLE
Adafruit_Neopixel library: RP2040 support added

### DISCUSSION THREAD

**Discussion by Bodmer** [earlephilhower/arduino-pico#98] [General]
I have adapted the Adafruit_Neopixel library to support the RP2040 using the PIO features of the Pico/RP2040.

[Download from here.](https://github.com/Bodmer/Adafruit_NeoPixel)

**Comment by earlephilhower** [comment]
Cool.  I have never actually used a neopixel but they seem to be everywhere!

Can you include the uncompiled source for the PIO.h file, and also look at the `PioProgram` class in `RP2040.h` for a simple way of working w/in the existing Arduino setup.  `Tone.cpp` or `Servo.cpp` shows how to use that class to work together w/others, and ensures only 1 copy is loaded in PIO memory.

You should also consider doing a PR on the docs/readme to include it in the list.

**Reply by Bodmer** [reply]
Done, apart from docs/readme.

**Reply by Bodmer** [reply]
Ah, had to revert, I was not testing the copy I thought I was!

**Reply by Bodmer** [reply]
Because the extension to the library follows the format of others (extern C) it cannot see the PioProgram class, so some restructuring will be needed.

**Reply by earlephilhower** [reply]
You can make the file `.cpp` and then just do `extern "C"  { ..... };` to allow C++ features but keep the C interface.  We do that in many files where the C++ compiler is used for better error detection/etc.

**Reply by Bodmer** [reply]
OK, will do it that way.

**Comment by Bodmer** [comment]
Adafruit have accepted a pull request on the current version so the official Adafruit_Neopixel Github copy of the library now has RP2040/Pico support built in.

Todo: add PIO availability check.
